### ISSUE TITLE
Add margins and padding to Control nodes

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#812]
**Describe the project you are working on:**
Space ship game that requires gamepad and keyboard input.

**Describe the problem or limitation you are having in your project:**
I keep running into the problem where I have to modify my theme or style or having to create additional padding nodes to position controls within a container.

I have tried using things like margin containers but this doesn't always work and requires a extra node just to position the control

these things start to add up and in no time I have 3 layers of containers to get the flow, position, size and clip I need for only one control.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I suggest allowing us to specify a margin and padding for controls that is separate from its rect.

this allows us to make gui easier since we don't have to search through a bunch of container children

I would also suggest that like the anchor a padding and margin rect be added to the godot editor to ease of use.
 
![image](https://user-images.githubusercontent.com/2453707/81457291-54879300-914a-11ea-8a17-7a1be615efde.png)

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
basically it would be similar to winforms but alot less buggy in which containers still position and resize their controls but users are allowed to adjust how they do it.

![image](https://user-images.githubusercontent.com/2453707/81457848-a29d9600-914c-11ea-9ef3-8ab40f02e34d.png)

 
**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
this enhancement would be use alot more then the current system. while it allows for bypassing a bunch of nested container controls it they would also still be useful for alignment.

I don't think this can be worked around due to how containers position their controls but if it can be it would require quire a bit of knowledge about the godot engine and some people just want to make a gui.

**Is there a reason why this should be core and not an add-on in the asset library?:**
this would be added in the core because currently the gui system can be bit infuriating to work with

**Comment by aaronfranke** [comment]
This is possibly a duplicate of #728

**Comment by KoBeWi** [comment]
> I have tried using things like margin containers but this doesn't always work and requires a extra node just to position the control

Well, working with UI requires lots of containers sometimes. The worst part is that smart snapping gets weird when you put too much on the scene, otherwise it's not really bad.
btw if you use PanelContainer, you can use `content_margin` property of StyleBox to replace MarginContainer.

**Comment by Shadowblitz16** [comment]
@KoBeWi 
this requires a bunch of nested controls just to get something like a clipped border.

the point of this is so that we have somewhat of a way to organize our controls

@aaronfranke this is not the same thing as he is talking about merging containers and controls while I am talking about allowing us to specify a margin and a border between containers and the control.

currently the margin container doesn't even work it this way since its position and margin are tied to each other and the parent container resizes the position and size

**Comment by groud** [comment]
> currently the margin container doesn't even work it this way since its position and margin are tied to each other and the parent container resizes the position and size

The margins of a MarginContainer are defined as custom constants, they have nothing to do with the basic Control's margins. This is basically stated in the documentation: [https://docs.godotengine.org/en/stable/classes/class_margincontainer.html](https://docs.godotengine.org/en/stable/classes/class_margincontainer.html)

That being said, I agree the naming is a little bit confusing. That's why I wanted, at some point, to rename the Control's margins properties to "offset_left", "offset_right" and so on. So that people stop understanding them as they were like CSS margins.

**Comment by Shadowblitz16** [comment]
@groud shouldn't the offset be ignored by the parent container then.

the point I am trying to make is its not very easy if at all possible to offset position and size of a control in a container.

every other gui library out there lets you do this.
the only extra part of this suggestion is to allow people to set a padding as well.

**Comment by groud** [comment]
> @groud shouldn't the offset be ignored by the parent container then.

It should yes. It's just that visually in the editor those values get updated while they should not. That's why editing those values should not be possible.

> the point I am trying to make is its not very easy if at all possible to offset position and size of a control in a container.

They aren't meant for it. If you need precise control on the positioning, you need to use a normal control as a parent instead of a container. The margins and anchors system allow you more custom positioning. And if the problem is only about the margins around the Control, then a MarginContainer works perfectly.

> every other gui library out there lets you do this.

Could you provide examples ? I am not sure a lot of libraries allow you to specify a padding for each Control/Container node, but I might be wrong.

**Comment by Shadowblitz16** [comment]
- winforms
- wpf
- any sort of wpf like library
- css


> > @groud shouldn't the offset be ignored by the parent container then.
> 
> It should yes. It's just that visually in the editor those values get updated while they should not. That's why editing those values should not be possible.
> 
they don't just get updated. position and margin are synced and margin isn't really a margin its a offset like you said

> > the point I am trying to make is its not very easy if at all possible to offset position and size of a control in a container.
> 
> They aren't meant for it. If you need precise control on the positioning, you need to use a normal control as a parent instead of a container. The margins and anchors system allow you more custom positioning. And if the problem is only about the margins around the Control, then a MarginContainer works perfectly.
> 

no containers are for layouts
controls should be positioned by these layouts but should also have say in the amount of spacing around them.

> > every other gui library out there lets you do this.
> 
> Could you provide examples ? I am not sure a lot of libraries allow you to specify a padding for each Control/Container node, but I might be wrong.
winforms
wpf
any sort of wpf like ui
css

you should really read up on the importance of margins and padding in ui.
I feel you don't understand what they are and how they help

**Comment by groud** [comment]
> they don't just get updated. position and margin are synced and margin isn't really a margin its a offset like you said

What I meant by that is that, as soon as you resize the parent Container, the position, size, margins and anchors properties are going to be overridden. They are consequently not reliable and should not be updated (or synced, as you wish), when modified in the editor. They should not be editable at all.

> no containers are for layouts
controls should be positioned by these layouts but should also have say in the amount of spacing around them.

No, that's not how Godot works. Once again, you want the engine adapt to your vision of things, but this spacing can completely be handled by a MarginContainer, as I said above.

> you should really read up on the importance of margins and padding in ui.
I feel you don't understand what they are and how they help

I already know what they are and I they work, but, once again, you did not provide any use case for them which are not already covered by another solution in the engine (you can use MarginContainer or even a normal control node in the hierarchy to achieve the same result). 

The only problem it could have with it is that it adds another node in your hierarchy, but honestly, I don't think it is a problem personally, as GUI layouts will always be complex anyway.

If we absolutely want to replace this intermediate MarginContainer by properties, I fear this will confuse users even more as they are going to be yet another set of Margins-like properties. This padding/margin will be used only inside containers as they can't be mixed with the current margins (or offset if we rename them), as those can be negative too, unlike padding. In the end, you node's position could be influenced by:
- anchors,
- margins,
- position,
- size,
- you new padding property

IMHO, the properties system is already complex enough, there is no need to complexify it even more.

**Comment by aaronfranke** [comment]
@Shadowblitz16 I would like to point out that you have the option to implement your own UI system in user code, using Godot's existing nodes as a base. You could create this as a proof-of-concept that people can try, or you can maintain it and publish it on the asset library.

**Comment by Shadowblitz16** [comment]
> > they don't just get updated. position and margin are synced and margin isn't really a margin its a offset like you said
> 
> What I meant by that is that, as soon as you resize the parent Container, the position, size, margins and anchors properties are going to be overridden. They are consequently not reliable and should not be updated (or synced, as you wish), when modified in the editor. They should not be editable at all.
> 

then this should be considered a bug if they are not stable enough to expose to the inspector. and in that case the gui system should be rewritten

> > no containers are for layouts
> > controls should be positioned by these layouts but should also have say in the amount of spacing around them.
> 
> No, that's not how Godot works. Once again, you want the engine adapt to your vision of things, but this spacing can completely be handled by a MarginContainer, as I said above.

godot doesn't work like that because its not implemented yet.
my vision is to make godot easier to use.
and we shouldn't have to add 50 margin container's just because we want to have a offset from the parent container.

also I have said this already margin containers don't even work for things like tab containers since the tab container resizes the margin container making the margin container useless.
I have already posted this issue which YOU closed that had a example project https://github.com/godotengine/godot/issues/38576

> 
> > you should really read up on the importance of margins and padding in ui.
> > I feel you don't understand what they are and how they help
> 
> I already know what they are and I they work, but, once again, you did not provide any use case for them which are not already covered by another solution in the engine (you can use MarginContainer or even a normal control node in the hierarchy to achieve the same result).
> 
> The only problem it could have with it is that it adds another node in your hierarchy, but honestly, I don't think it is a problem personally, as GUI layouts will always be complex anyway.
> 
> If we absolutely want to replace this intermediate MarginContainer by properties, I fear this will confuse users even more as they are going to be yet another set of Margins-like properties. This padding/margin will be used only inside containers as they can't be mixed with the current margins (or offset if we rename them), as those can be negative too, unlike padding. In the end, you node's position could be influenced by:
> 
> * anchors,
> * margins,
> * position,
> * size,
> * you new padding property
> 
> IMHO, the properties system is already complex enough, there is no need to complexify it even more.

I provided at least 2 reasons why this would make the gui system better why don't you reread this suggestion.

I don't see how you don't think padding is a good thing to add to controls.
by your logic why do we even have anchors and margins if they can just be another node.
controls could simply just be a position and size but they aren't because it would again generate too many node to do anything.

this isn't any more complex then any other gui system.
they all have..
- anchors
- padding
- margins
- a rect
 some even have a border which is a one extra thing

**Comment by trommlbomml** [comment]
I can follow @Shadowblitz16 argumentation and I would like to add some of my point of view which are often needed - especially if you rely heavily on layouts. Of course you can solve it differently with godot, but this is such a fundamental concept I really would like to take into consideration of redesigning the already great UI solution.

I also would like to point out "make your own solution" sounds like there are arguing points which are good but are not taking into consideration as this would change the UI system to much that backard compatibility is really hard?

## Padding for Button content
the text and the distance to the border of the button should easily set up using padding. Often it looks way better to have some horizontal and vertical space between text and border. Having a property for that would be fantastic.

## Margins for any control improves usability to improve UI quality
The current margin seems not be a layout but a relative positioning concept. Margin is "normally" an additional spacing around the control, the godot margin seems to inset that - thats really weird when you come from anything else. It is a common use case to tightly pack the elements using layouters and then add margins to controls to make it look good - and it is a fast and reliable workflow: You can concentrate on layout - as you would expect from a layouter as the name says - and define the distances with margins in a separate step simply by modifying the controls margin. My approach is currently to add everywhere margin containers where it is needed, this takes a lot of time and I would be much faster using a typical margin system.

## Technical side
The implementation should be simple and can be integrated in control base class by simply adding the margin during size calculation and remove it during rendering phase and the benefit would be the system is much clearer and nearer to what anybody knows from one of the most common techniques for styles and layouting: CSS. 

##Downsides
As already said: backwards compatibility? People who like the current system and are not willing to change their mind?

**Comment by Kiaazad** [comment]
I'm coming from a renpy/unity background as a GUI designer. I've heard lots of people praising Godot's versatile interface design tools, but as soon as I've started using it, I've found the way Godot handles interface cumbersome.

The biggest advantage of adding margins, alignment and padding directly to the containers and buttons is simplicity. This simplicity removes the barrier that stands in front of developers migrating from other platforms.
There's also an obvious benefit to mimicking other successful platforms: flattening the learning curve.
Unity is switching to CSS for it's interface design by introducing their "UI builder", and that move alone makes it the engine of choice for tons of web developers that might have an idea to build a game or app, growing their user base by a huge margin.

I don't know why people argue against improving the tools they're using. It's not like we're asking to remove those center, align or margin containers and break backwards compatibility, all we ask for is improving the rest of them to make it easier for the newcomers and not leaving them scratching their heads for something that was simple everywhere else.

**Comment by Calinou** [comment]
> I don't know why people argue against improving the tools they're using. It's not like we're asking to remove those center, align or margin containers and break backwards compatibility, all we ask for is improving the rest of them to make it easier for the newcomers and not leaving them scratching their heads for something that was simple everywhere else.

Adding multiple ways to achieve the same thing is not necessarily a net upgrade in terms of usability. It can introduce choice paralysis while making it more difficult for maintainers to fix bugs in the GUI system.

**Comment by Kiaazad** [comment]
> Adding multiple ways to achieve the same thing is not necessarily a net upgrade in terms of usability. It can introduce choice paralysis while making it more difficult for maintainers to fix bugs in the GUI system.

Adding multiple ways is the first step of letting the users decide which way they prefer, specially when one involves much less work. We can deprecate the one less popular to prevent the slim chance that somebody couldn't decide which one to use.
I should warn you about cult mentality when you do something a certain way and believe it's the best till you get used to the way things are and resist any change. It's not that hard to find an excuse to maintain the status quo if you are looking for it.
I would love to be an advocate for the Godot engine, but it had to shoot that arrow to my knee and send me searching fora solution to some problem that shouldn't exist in the first place. instead of pushing my clients to switch to Godot, I'm arguing on Github for something that's clearly bothering the users to the point that there are multiple issues submitted for it. 
all I'm saying is: Give us the legs to walk on our own before preaching about the efficiency of wheels.

**Comment by groud** [comment]
That won't happen sorry. We won't maintain several systems if the use cases are the same, especially if the only reason to have another system is "for the sake of letting people have the choice". As @Calinou mentioned, the additional workload on contributors would not be worth.

However, you are still free to create a new GUI system via a plugin, maintain it, then make it used by a significant amount of people to proove us your solution is better than the current one. In such case we would consider integrating it, or even replacing the current implementation. But otherwise, this won't happen.

**Comment by Kiaazad** [comment]
I would do that if I was using Godot for long enough to know what goes where. But you can't expect that from a newcomer, and those experienced enough to do so, are already part of the `no change necessary` cult.
I'm sure you can see the flaw in in your otherwise perfectly reasonable suggestion.

**Comment by groud** [comment]
> and those experienced enough to do so, are already part of the no change necessary cult.

Well,  wouldn't say it's a cult... And I would even say it's actually the opposite, since we are asking for good rational reasons to change our system, while you're advocating for changes for the sake of it. You know, " you should look at the beam your own eye instead of the mote in ours." ;)

**Comment by Kiaazad** [comment]
I won't argue what is, and what isn't a cult mentality. Instead...
I'll suggest: instead of rejecting the idea, point out the people that are asking for the feature to the proper starting point to implement what they are asking for, perhaps as a plugin that you've suggested.
And since it's the issue that's frequently coming up, maybe assign an experienced developer to guide them.

**Comment by groud** [comment]
> I'll suggest: instead of rejecting the idea, point out the people that are asking for the feature to the proper starting point to implement what they are asking for, perhaps as a plugin that you've suggested.
And since it's the issue that's frequently coming up, maybe assign an experienced developer to guide them.

I'd be happy to help working on a solution, if we identify where the problem is, and if it worth solving because everyone would agree it's a problem. Right now, the only problem I identified in this proposal is the fact that our system leads to big nested structures of nodes. For me, it's not a big problem, and so far, we don't have a lot of complains about it too. So consequently, there's little interest in experienced contributors working on a solution for this (directly or indirectly). Note that you are welcome to prove me wrong, if you gather a lot of people having a big problem with this situation, we may reconsider the importance of the problem.

You can have a look to our workflow, it explains how we select things to implement pretty well: https://docs.godotengine.org/en/latest/community/contributing/best_practices_for_engine_contributors.html

**Comment by KoBeWi** [comment]
> And since it's the issue that's frequently coming up

Is it really that frequent? Only 3 people commented here that aren't Godot engine members and the OP has no upvotes. Personally I don't really recall this problem being mentioned elsewhere (although I don't follow every community channel much). I mean, it's not really apparent that this proposal has much support.

As for the implementation, doing it via plugin would probably involve writing new GUI system. It would be easier to make a PR and maybe distribute a custom build for interested people. All relevant code to implement this proposal should be inside `control.cpp`.

**Comment by trommlbomml** [comment]
I am also a bit astonished about how people argue against something which is well established solution in UI systems - like e.g. the anchor system. Furthermore this is an approval issue tracker and I think that's also a reason there are not so much upvotes? 

@KoBeWi it is not frequent because people are told e.g. add a margin container, update the texture or simply increase the size, that's also what I think @Calinou means there are other ways to do it - which is I think not ideal or basic problems like making layouting simple.

I also think that a lot of UI designs in games are more based on mockups and not on layouts like you do for business applications, where you need way more flexibility than in a compared simple game UI.

I think a good Idea would be using a pull request, too. At least my intention was to produce some focus on this topic to make more people aware of this property.

**Comment by Kiaazad** [comment]
It's a case of "Abraham Wald and the Missing Bullet Holes". Most of those whom have this problem just give up and move unto another engine. The few that bother to submit an issue, really care about open source and the potential value that Godot can bring to their workflow.
In my case, I've already developed the rest of the game for my client, moving onto another engine means starting from scratch and I'm too lazy to give up that easily.
The only argument I keep hearing is: `For me, it's not a big problem`. In my opinion that is the wrong attitude towards those whom see it as a big problem and try to help correcting it.
I've pointed out several problems with the current system that you've already agree on at least one, but let me list them again in case I didn't get any of those points across:

1. The current system leads to big nested structures of nodes.
2. It's very unintuitive for those whom are migrating from other engines.
3. It's very unintuitive for those with a background in some of the most popular languages out there, (don't tell me 94% of the internet is wrong with their fancy CSS styles)
4. The learning curve is too steep.
5. We're losing new users to maintain the status quo.
6. My clients are not happy about the slow progress of their project (yes I'm counting that as a problem)

**Comment by Zireael07** [comment]
@KoBeWi: IIRC we had a similar issue back on the bug tracker and it had more support there. 

The main reasoning in favor was familiarity for users of most engines out there (and yes, that includes CSS)

**Comment by bluenote10** [comment]
> Most of those whom have this problem just give up and move unto another engine.

Fwiw, I also came with a CSS background, got confused by Godot's notion of margins, and had a similar initial impression as you. However after using it for a while the system started to make sense, and perhaps it is even superior to how CSS works.

**Comment by Calinou** [comment]
@bluenote10 Renaming margins to offsets might be enough to cure the confusion :slightly_smiling_face:

**Comment by trommlbomml** [comment]
I also hope CSS like support is not coming to Godot as a core features - it's hell in detail to layout everything good, other layout systems are way easier including Godot. But having the margin model and maybe call it different would be a good solution I think. Maybe we can refer for this not as offset but maybe spacing as it is indeed adding space around controls?

**Comment by Shadowblitz16** [comment]
the issue is not the naming its that there is no way to specify outer spacing
The containers resize their children which have no control over their size forcing the user to have a extra node for every container in  the scene.

**Comment by KoBeWi** [comment]
> forcing the user to have a extra node for every container in the scene.

Why exactly an extra node is a problem?

**Comment by Shadowblitz16** [comment]
@KoBeWi  same reason an extra 1000 nodes is a problem
- it takes up space making the scene view harder to navigate
- it allocates more data making gamer bigger and slower
- its unnecessary and is the reason why padding, margins, and borders exist in the first place

**Comment by trommlbomml** [comment]
> Why exactly an extra node is a problem?

Already mentioned in the thread by @Kiaazad: if you need it a lot the tree grows massively just because you want to have some space between elements, this is from a usability point very bad. I mean the property for horizontal fill/shrink start/end are also not nodes but properties of control.

**Comment by Kiaazad** [comment]
Here's how I make a frame with 3 buttons in renpy.
```python
frame:
    has hbox
    label _("Rollback Side")
    textbutton _("Disable")
    textbutton _("Left")
    textbutton _("Right")
```
and it looks like this: 
![image](https://user-images.githubusercontent.com/33455007/93304567-0ea2a180-f7b2-11ea-9111-31b01f4adfb3.png)

Can somebody show us how a superior system does it please.

**Comment by trommlbomml** [comment]
To even support @Kiaazad what I mean: 

1. I do not want to change the ninepatch texture to have the distance from the text to the button, padding/margin would solve this easily
2. I do not want to add a margin container to a vbox container to the label, the box and the panelcontainer, I have to think about this building the core layout and adding it afterwards is really clunky.
3. I do not want to not use a button and replace it with a texture rect and manually sizing just because I can do it.

simple would be 5 nodes:

- panel container
    - vbox
        - label
        - button
        - button
        - button

and add for each element just margins on the controls itself. the existing solution would lead to the following layout:

- panel container
   - margin container
    - vbox
        - margin container
            - label
        - button
        - button
        - button

And the point is this is a small, simple UI. if you have larger UIs with more cases you really mess up your already quite complex layout with a lot of margin container nodes which, makes it harder to read. And I think one of the core good concepts of godot is that the tree represents an easy view of your game.

**Comment by Kiaazad** [comment]
Well, @trommlbomml, I was waiting for one of the defenders of the current system to answer:
![image](https://user-images.githubusercontent.com/33455007/93312470-1c5d2480-f7bc-11ea-8bee-fc651eaf03df.png)
Then I would unleash my gotcha moment to ask:
VVV
Very good, It feels great when the teacher asks you an easy question doesn't it?
Now do this:
![image](https://user-images.githubusercontent.com/33455007/93312633-4dd5f000-f7bc-11ea-8c03-61438aab5a02.png)

It was as simple as:
```python
frame:
    xysize 700,300
    has hbox
    yalign 0.2
    label _("Rollback Side")
    textbutton _("Disable")
    textbutton _("Left")
    textbutton _("Right")
```
And for my next trick: I would simply ask them to resize the frame without the whole interface going through a mental breakdown. But I guess everybody have seen it coming from miles away.
I'm sure after enough fiddling with the numbers, everything is doable in the current system, but that fiddling is costing us our time and sanity.

**Comment by trommlbomml** [comment]
I think your point is a bit biased to a solution of a different system you like, we should focus on arguing how this can be simpler with the provided proposal in this issue in Godot.

**Comment by silentorb** [comment]
Over the years I've increasingly gravitated toward minimal box models where each UI element does one thing and does it well.  This leads to more flexible layouts that are easier to reason about, debug, and more performant.

This is because the more UI systems rely on kitchen-sink widgets, defining layout becomes less about addition and more about subtraction, where the developer must fight against an array of gravitational forces.  And when elements become heavier, there becomes less room for lighter, more specialized elements.

If anything, I'd like Godot's base Controls to be leaner.

One of the concerns raised in this thread is node complexity.  That is not really a matter of layout, but of abstraction.  The solution to complexity is rarely to globally make each element more complex, it's to break up the problem into smaller, more manageable pieces.

The real question is whether Godot's UI system supports abstractions powerful enough to resolve the problems raised in this thread.  If not then the UI may need better abstraction abilities.

As an example, creating disciplined web styling means rarely EVER specifying margin and padding on a particular element.  Spacing should already be handled by consistent theming.  (That's just an example—I'm not advocating CSS.)

**Comment by d3is** [comment]
As a Front-end develop of over 15 years I have to fervently disagree with your idea of disciplined "web styling" never using Margins and Padding. With things such as VH and VW in CSS, we have such an extreme level of control for responsive layouts that I believe it is irresponsible and restrictive to not utilize Margins and Padding responsibly. If we had something as simple as how VH/VW work in CSS, then we would skyrocket above every other Game engine's UI system out there. I actually hope I am just misunderstanding what you said, because I am very confused how you could not take advantage of both Margin and Padding.

I truly appreciate your desire to keep things lean, as I have over the years always preferred to use as little as possible when it comes to front-end frameworks, etc. This is why I use Godot over Unity now. I have leaned on Eric Meyer's CSS reset for my entire career. But, not having something as simple as Padding in a UI is just plain silly. For Godot it doesn't even have to be applied to multiple containers, we simply need a PaddingContainer for now. Ideally in the future we take advantage of systems like a VH/VW from CSS that factors in the width/height ratios.

**Comment by KoBeWi** [comment]
> we simply need a PaddingContainer for now

How is it different than MarginContainer?

**Comment by d3is** [comment]
A MarginContainer's Left/Top/Right/Bottom values do not consider the child elements sizes.

**Comment by aaronfranke** [comment]
@d3is A MarginContainer, with its child set to "Fill" on both directions, already does this. The child element will fill the width/height minus the margin and must fit in this space. Maybe I'm not understanding correctly, can you provide an example?

**Comment by d3is** [comment]
If you have a Margin or HBox container with 2 or more Children, you cannot let the children expand to a Max size of the parent container with a padding around them that will wrap any following children to the next line. This actually outlines 2 problems in that you cannot set a padding, or a max size for a basic container.

**Comment by KoBeWi** [comment]
> you cannot let the children expand to a Max size of the parent container with a padding around them that will wrap any following children to the next line

If it's about wrapping, then it sounds like a missing Container type. Instead of adding properties to every Control node, we could just add a WrapContainer (or whatever it is called) that would try to fit its children without expanding. This is totally doable and doesn't require any changes to the current API.

**Comment by Kiaazad** [comment]
before rambling about how godot interface tools don't make sense, I have to ask: how do you even make a text inside a panel with some margin around the text. shouldn't it be something like:
![image](https://user-images.githubusercontent.com/33455007/97665485-c0332480-1a38-11eb-8a17-fca23b146487.png)
maybe I'm dumb or something but this is an example of what I mean by unintuitive.

**Comment by aaronfranke** [comment]
@Kiaazad Godot's UI system is mostly outside-in, you have to make the parents bigger first, and the children don't get to resize their parents. In particular, containers such as MarginContainer determine the size and position of their children with minimal exceptions, such as the minimum size property, and Label resizing itself (but not its parents) to fit the text.

**Comment by Kiaazad** [comment]
> 
> 
> @Kiaazad Godot's UI system is mostly outside-in.
That's a major flaw, and I think the source of every problem mentioned in this topic.
When it comes to advanced interface design, the content is the king and any container have to scale and fit the king, not the other way around. a dialogue box needs to accommodate the dialogue, big or small, nobody likes to have limitations on how much text they can show.
I did suspect that is the case and was about to submit a ticket to change the interface elements design to inside out, but it was so unlikely that somebody made this mistake and nobody have noticed it yet, I had to ask first.
If none of the developers are willing to spend their time to fix this flat tire, please provide some information where I can start and I'll somehow do it myself.

**Comment by aaronfranke** [comment]
@Kiaazad I personally much prefer Control nodes as they are to HTML/CSS, and I think many/most Godot users would agree. I doubt Control nodes will fundamentally change in this way. Control nodes are battle tested and work great for many types of UIs without dynamically sized content, it works great for menus of games, and don't forget the Godot editor itself.

However, you should be able to use Godot's Control nodes as a base (or maybe CanvasItem) and design your own UI system. It would be a great idea for an add-on in the asset library for people that want content-aware inside-out UIs. If such a system were optional and could exist in parallel with Control nodes, I would imagine it would be received quite positively.

**Comment by Kiaazad** [comment]
> Control nodes are battle tested and work great for many types of UIs without dynamically sized content,

that's the exact problem, once an engine is past the flappy bird type of games, bigger game developers come in and they have a vision of how the information must be shown, it wouldn't be about showing coin count and a single health bar anymore, they expect to add any weird item, spell or quest they have dreamt last night and the interface must accommodate however size the name and description is, `great for many types of UIs without dynamically sized content` doesn't cut it anymore.
can somebody give me a link to some documentation page that contains the instruction on how to change/extend the nodes to create the correct type of nodes please. don't assume everybody have spend as much time as you have on godot and automatically knows everything you know when you mention add-on, asset library and control nodes. I've only spent few hours on godot before I got frustrated and closed it. pretend you're talking to a complete noob.

**Comment by aaronfranke** [comment]
@Kiaazad Designing a whole new UI system is a complex task. If you wish to do this, I highly recommend you familiarize yourself with Godot first, including how the current UI system works and what it can/can't do.

But briefly: [Control](https://docs.godotengine.org/en/stable/classes/class_control.html) nodes are the base type for [Godot's UI system](https://docs.godotengine.org/en/latest/tutorials/ui/index.html). All UI nodes inherit from Control, and they are all colored **green**. Control inherits from [CanvasItem](https://docs.godotengine.org/en/stable/classes/class_canvasitem.html), and CanvasItem inherits from [Node](https://docs.godotengine.org/en/stable/classes/class_node.html). CanvasItem is the base for all 2D nodes, and only Control and [Node2D](https://docs.godotengine.org/en/stable/classes/class_node2d.html) (the **blue** ones) directly inherit from it.

[An add-on (aka plugin)](https://docs.godotengine.org/en/stable/tutorials/plugins/editor/making_plugins.html) is a folder that contains a scripts/scenes/whatever and also a config script. Add-ons can be distributed just by copying the folder, or via the [Asset Library](https://godotengine.org/asset-library/asset) which allows users to download and install add-ons from inside the editor.

**Comment by Kiaazad** [comment]
> @Kiaazad Designing a whole new UI system is a complex task. If you wish to do this, I highly recommend you familiarize yourself with Godot first, including how the current UI system works and what it can/can't do.

I design/code UI systems for a living, it's not that complex to me and it's not an issue, my whole problem is what the current UI system can't do. it should, but can't. that's all.

thanks for the links and description, much appreciated.

**Comment by voidshine** [comment]
As a newcomer from a mixed background (Web, WPF, Unity, many others...), I quite like Godot's layout system. It's simple, composable, and gives direct control. Like any system, you have to learn its ways and quirks, but once you know how a margin container works, it becomes a versatile tool in your kit. You can use it for margins *or* padding, for example; or you could script it to push elements apart very predictably.

CSS is a different beast; it's powerful and yields quick results but, in my experience, very complex and often surprising. Working with it feels less like programming and more like giving suggestions, hinting and hoping. Still, HTML & CSS are ubiquitous for a reason. Sometimes you just want to throw a bunch of content into a DOM and tweak styles until it looks decent. I don't think Godot should go that way; its simplicity is its strength, and HTML/CSS are anything but simple. Totally +1 to the plugin idea, though.

Unity's latest UI system is going with Facebook's Yoga layout library. Anyone could clone [the Yoga repository](https://github.com/facebook/yoga) and drive Godot controls with its flexbox layout implementation. Happy coding. :)

**Comment by Shadowblitz16** [comment]
I don't think forcing a dynamic layouts is a good thing which is why I hate wpf and love winforms.
but godot's ui has some serious flaws in several places that I have run into every time I use it.

I think godot ui would be much easier if it was inside-out.
If you don't want your controls to resize their parent then don't set their size flags to expand or fill

**Comment by trommlbomml** [comment]
> 
> CSS is a different beast;

I am at least do not want to have CSS-Like layouting in godot - PLEASE NOT! it's hell in detail, especially if you are in the browser. The original discussion was about that margin is a property of a control like their positioning, sizing behavior and anchoring to the parent. You could also argue that anchoring should be a separate control as you may not needed it inside a layouted container - thats the same for the margin/padding and the core discussion. We know that you can use the node, but it is very verbose and blows up the tree - which should show the structure - and margin is just a small modification of the control and not a dedicated one.

> I don't think forcing a dynamic layouts is a good thing which is why I hate wpf and love winforms.

I think thats not what godot forces you, simply do not use anchors and layout by coordinates manually as in winforms. I think that is also a valid solution for fixed resolution/pixel-perfect games.

**Comment by Shadowblitz16** [comment]
@trommlbomml the issue of this was originally to add padding and margins like in winfroms
we don't need borders just a way to offset controls in containers.

also If you have seen my newer issues I don't want to force dynamic layout on people but right now some containers do not take in account their children's size flags. which mean it crushes them down to their minimum size regardless if we want them to or not.

**Comment by Kangrew** [comment]
Having margin, padding and border would make it so much more easier and simple especially for someone like me who previously struggled with godots Container approach.

**Comment by pycbouh** [comment]
Margins are a part of the positioning, while padding and border are parts of the visuals and internal logic, and are handled by styleboxes. So it's already possible to configure these for individual controls.

**Comment by Shadowblitz16** [comment]
@pycbouh
Then let us add a offset to the border and padding

**Comment by pycbouh** [comment]
You already can. Use a stylebox on controls that support it, or make your own controls that draw a stylebox and then use it as well.

**Comment by Shadowblitz16** [comment]
no I mean offset it through the control not the stylebox

**Comment by pycbouh** [comment]
Why complicate things and separate related parts into different elements if they need to be ultimately considered by each other? Makes no sense. There is a tool for this in Godot's UI system. Use it.

**Comment by Shadowblitz16** [comment]
@pycbouh you say it complicates things I say it simplifies them.

people can then separate the default control layout and the theme it uses.

**Comment by pycbouh** [comment]
There is no default control layout. And border is not a part of the layout, it's a part of styling, theming. So is inner margin/padding. This is not CSS where those things mix up.

**Comment by Shadowblitz16** [comment]
The default layout is simple.

margin  = The layout of the control relative to the parent.
padding = The border around the control

Not sure what border is. I never used in in winforms.

**Comment by pycbouh** [comment]
I know what you mean, but this doesn't exist in Godot, because controls are not following the box model or anything similar. There is a different approach in Godot's UI system, and it works well.

**Comment by Shadowblitz16** [comment]
@pycbouh is disagree its confusing, limited and has inconsistencies when it comes to the layout of parents and children.

**Comment by Kangrew** [comment]
I agree with @Shadowblitz16 , the naming is really confusing and I often get things mixed up. I think that is at least something that Godot has to change and come up with different names to that used in the css box model as it easily misleads someone with a web or similar background. I refused to use Godots container approach for a very long time because of this.

**Comment by pycbouh** [comment]
@Kangrew There is already a proposal to rename margins to paddings where appropriate.

**Comment by Kangrew** [comment]
I was suggesting something completely different to box model names. I think people new users will start expecting it to work like css.

**Comment by pycbouh** [comment]
Those words are not exclusive to CSS and box model. We shouldn't go out of our way to distinguish from that, their meaning is pretty clear and used for various layout systems and just visual design in general.

**Comment by Kiaazad** [comment]
I'd say add margin, padding, align and offset to all container, and keep the container form of them the way they are for anybody whom like it that way.
It's clearly an issue bothering the newcomers and addressing it can't negatively effect the existing users in any way, just don't use them. The thing I'm baffled by is the amount of resistance towards this change.

**Comment by pycbouh** [comment]
> I'd say add margin, padding, align and offset to all container, and keep the container form of them the way they are for anybody whom like it that way.

You already have margins for controls and containers that can be positioned with margins. But there are other methods of positioning which do not involve those. See https://github.com/godotengine/godot-proposals/issues/2841 for a discussion on what modes of positioning are available, when they are applicable and how to make them more clear to the users.

Paddings is what `MarginContainer` is for, but I personally wouldn't mind it being a property rather than a node. I think having it as a separate node allows to simplify the drawing logic for every control, as it only needs to worry about its own rect size and nothing else. This can probably be solved by helper functions, but the change won't be easy and will have a cascading effect. Having a node isn't that bad anyway, but I agree with https://github.com/godotengine/godot-proposals/issues/2334 that it can be renamed to PaddingContainer, so that its properties are not confused with margins used for positioning.

There is no universal rule for alignment. Some containers do have alignment as their inherent property but that ties to their special behavior. Overall alignment strategy depends on which positioning mode is applicable, and can be related to anchors, grow directions or some container-specific behavior. Once again, see https://github.com/godotengine/godot-proposals/issues/2841.

I don't know what you mean by offset, because it's just a generic word and both margins and paddings are offsets.

> It's clearly an issue bothering the newcomers

There is no universal truth to it. Yes, people coming from using a different UI system would need to learn this one. Like I've said, we can and should work on the presentation of data, but the mere fact that there is different kind of system used for UI in Godot doesn't mean that it's a bad system. And the fact that it requires mastering doesn't mean it's unworthy of being mastered and needs to be replaced with some other run-of-the-mill approach.

> addressing it can't negatively effect the existing users in any way, just don't use them

Most proposals regarding changing the way our system works are based on misunderstanding the system, its compromises and its patterns. And so they often ask for unrealistic "just add this for us, don't use it if you don't need it" when in reality those changes would have great affect, or would introduce an unnecessary split in user-land approaches to solving problems.

I firmly believe that this is more of a UX problem than the underlying implementation problem. If Godot had, say, a UI designer, nobody would care for the internals which currently are exposed more or less as is. Same as nobody would suggest we move to the box model or implement CSS support if they new how awful those things are internally and how many hacks they require to function in a sensible way.

**Comment by Kiaazad** [comment]
I've seen those same arguments before. It's the resistance I'm talking about.

> There is no universal truth to it.
I don't know what passes as a universal truth, but when people bother to come to github and complain, it means they have exhausted every other option they could try and their problem persists, reading the documentations, watching videos, searching google and asking on the forums. Now imagine how many gave up on the way.

> There is no universal rule for alignment.
I don't think anybody would write a universal set of rules for something like this. usually a node can be aligned to some fraction of it's parent, on x and y axis, be it another node or the screen. it's not that complex, if there is extra room, the node can be on one of the sides or somewhere in between. We can just look at how others do it and learn from them.

> Most proposals regarding changing the way our system works are based on misunderstanding the system.
If the system causes lots of misunderstanding, maybe there's something wrong with the system.

Even if none of suggested solutions where valid, the fact that there's a discussion on the topic means there's a problem. You can't argue away the source of people's dissatisfaction with the status quo. It's not the problem that go away with these arguments, it's the people.
I, as a sample of one, decided to migrate to godot years ago, got into problem with the interface design tools, spent doing all of the aforementioned steps. gave up several times, last year decided to try harder, landed on github discussions and got chased away by the same arguments that I'm still hearing. I know the next arguments would be along the lines: godot is open source, do it yourself. well... I would if it was along my area of expertise. I can offer my years of experience in UI/UX design instead. All I'm asking in return is to be heard.

**Comment by pycbouh** [comment]
> I don't know what passes as a universal truth, but when people bother to come to github and complain, it means they have exhausted every other option they could try and their problem persists, reading the documentations, watching videos, searching google and asking on the forums. Now imagine how many gave up on the way.

And when they don't have a problem they don't come to GitHub and don't express how it just clicks with them or how they've mastered it after a while. This metric is not representative, it's a vocal minority bias.

> Usually a node can be aligned to some fraction of it's parent, on x and y axis, be it another node or the screen.

If you don't ignore the rest of my explanation, you will learn that the "universal rule" was in relation to different types of containers. You ask us to put some "align" property on the base container. This is impossible, because there is no one way of alignment implemented — this is what different ways of positioning and different types of containers are for. If you want to just align controls against other controls, you may not need containers at all; that's what anchors are for.

> If the system causes lots of misunderstanding, maybe there's something wrong with the system.

You see resistance, and I see baseless generalizations about "lots of" problems. Like I've said before, for every report from people misunderstanding the system there is a person who doesn't and therefore doesn't report anything to GitHub.

> Even if none of suggested solutions where valid, the fact that there's a discussion on the topic means there's a problem. You can't argue away the source of people's dissatisfaction with the status quo. It's not the problem that goes away with these arguments, it's the people.

And I don't deny there is a problem, look at my closing paragraph:

> I firmly believe that this is more of a UX problem than the underlying implementation problem.

**Comment by Kiaazad** [comment]
If others have done it and got better results, with less frustration for the users, maybe godot's approach isn't that optimal.
I don't see that many people having trouble with the interface tools that some of much smaller engines provide. It can't be the users either since I've tried many engines and only godot gave me this much grief. but neither are the main problem. THe main problem is the answers.
Many users come and suggest things with good intention, they see something missing or something that can be improved and mention it because they like everything else about the engine, but all they find is a wall in front of their face telling them no.

**Comment by pycbouh** [comment]
You seem to be frustrated and ignore a significant chunk of what is being explained here. I'm sorry that you've ended up feeling ignored, but please keep in mind that even when some of the maintainers tells "No" to your idea they also come with good intentions but probably a different outlook on the problem. Rejections is a part of any proposal discussion, please don't take it personally.

**Comment by Kiaazad** [comment]
It's more the reasons given are the thing that bother me than the rejections. Just because some of the users got used to the current system, all new users that have a difficult time with it are rejected. it's like comparing godot to other engines is a sin and forbidden.

**Comment by Killfrra** [comment]
I would suggest not to invent your own UI system, and instead, change how the current system is represented to the user.
I started writing a plugin for Godot 3.3 in GDScript, which would solve this issue and #728 as well, but faced with difficulties. It would be much easier if you could inherit, for example, from SceneTreeEditor, override its methods and have access to its variables, as it would be with a node written on GDScript. But since there is no such possibility, I had to do with frank hacking: to rebuild the Tree, intercept signals, temporarily push data to where it will later be read by the function, determine the TreeItem's button id by Texture... Still couldn't get CanvasEditor and Inspector to edit different nodes at the same time. Perhaps some of the problems are solved through the writing of the module in C++, but I am worried that the potential user will have to compile it. It seems to me that the lightweight engine that tries as much as possible the functional to be taken into plugins should be easier in the extension. But in any case, it is cool that the engine is built on the same basics as the games in it. If anyone wants to help with the writing or to see what I've done, you are welcome to [my repo](https://github.com/Killfrra/UISimplifier)

**Comment by Kiaazad** [comment]
I would suggest allowing the users to design their interface in another tool and add it to the game in something like uxml and add it on top [like how the new unity does it](https://youtu.be/t4tfgI1XvGs). Clean and modular.
Clean and modular. No need to touch the existing system either, just hook it into the engine and let the users decide, although, knowing the community I'm sure somebody will find something severely wrong with this idea and shoots it down in one swift excuse.

**Comment by pycbouh** [comment]
@Kiaazad You are completely free to make such a tool and integrate it with Godot's UI system. Nobody can shoot it down if that's what you want to do.

**Comment by Kiaazad** [comment]
> 
> 
> @Kiaazad You are completely free to make such a tool and integrate it with Godot's UI system. Nobody can shoot it down if that's what you want to do.

Ah, the "Godot is open source, do it yourself!" excuse right out of the bat.
Don't you think if I could, I would've already done it instead of coming here suggesting it?
Unfortunately as a new user, there is a big barrier in front of me, that is Godots interface tools, it's clunky and unintuitive, making it unbearable to learn.
first let people in then except them do these kind of things.

**Comment by pycbouh** [comment]
If you were proposing we create a dedicated tool that implements a separate, contesting system for the one we already have as a core project, like Unity does for a lot of its technologies, this will not happen. We have a limited amount people as it is, not enough to support two contesting stacks.

It was stated many times before that we are open to making UX changes to make the process of learning Godot's UI system. There are in fact a number of proposals and PRs targeting that.

As for calling it an "excuse", well, as an open source project we live and die with our community and our contributors. This is the matter of fact when picking Godot over something else, and it may not be for everyone. Many regular contributors work voluntarily and paid contributors are numbered. So indeed, if there is a need in community to design a completely new process for UI because for some learning Godot's system is a complicated task, it would be in the interest of the said community to start some work on it. Good work may end up in the engine and help lots of people, niche work can still help those who need it without overburdening the core engine.

**Comment by Kiaazad** [comment]
> We have a limited amount people as it is

I can guess why that is the case, I personally love Godot, sans it's interface tools of course, and as an experienced UI designer I offered to help improving it but I was denied at every door.
The amount I can contribute might be small to begin with, but you can't complain about not having enough people while closing your doors on everybody's face.
It's been a year or more since we have pointed out the problems with the interface nodes and suggested ways to remedy the situation, the best thing we got in return was being ignored if not denial of the said problems or refusing to listen to our reason.
Again, I'm extending my helping hand to anybody whom is interested in using my experience as a UI designer to improve the user experience when it comes to using the interface tools. I can't code but I can point out where it can be improved and how.
Let's watch it getting slapped away yet another time.

**Comment by Zireael07** [comment]
"It's been a year or more since we have pointed out the problems with the interface nodes and suggested ways to remedy the situation, the best thing we got in return was being ignored if not denial of the said problems or refusing to listen to our reason."

I don't see many thumbs-up on the proposal and I only see what, three people in favor of it in this proposal thread. 
No one is refusing to listen to your reasons (this proposal is still open) but I simply don't see as much of a support for this as you seem to think there is (see this nice website https://godot-proposals-viewer.github.io/ displaying proposals in a rough popularity order).

Most proposals, especially very complex ones, get done because there's clear, overwhelming support for them AND there is a contributor on hand who can do it. Again, just by glancing at the list linked, there are many proposals with overwhelming support that haven't happened yet because of the limited amount of contributors...

**Comment by pycbouh** [comment]
This was not a complaint, it is the reality of the situation. Same with issues taking long time to resolve sometimes. It is what it is.

I don't know who've closed the door in your face many times before but the entire time you and I are discussing things here you have been extremely negative towards everything said to you and very dismissive of what Godot is and what the arguments against your points are. This attitude doesn't help others to hear you out. It's clear that you don't like what is being explained here, but pointless bickering doesn't move the discussion forward.

I've already extended my regards about the way you feel you were treated before, but I can't do it every time to revive this thread to curse the developers for not seeing things the way you do.

**Comment by pycbouh** [comment]
And by the way, we have this discussion on a proposal that has exactly one vote in favor. After a year of being open it can be closed at this point as not desired by the community.

**Comment by Kiaazad** [comment]
Ouch... one slap was enough...
**First**: I'm not reviving this issue, every now and then I get an email notification that somebody commented on it and since not much have changed since the last time, I chip in with my disappointment. And I don't remember cursing.

**Second**: #2841 is basically pointing to this same issue of the interface tools being unintuitive, and it's one of the top issues on the nice page that Zireael07 presented. I'm sure there are many issues posted on something related to this that don't get that much attention like this one, but it doesn't mean none of them are valid.

**Third**: nobody cared enough to point us to the popular ones so we can support them to be done, that's another way people get ignored.

**And finally**: Thank you for letting me know how much my concern and offers (bickering) means to the community and the fact that they are pointless.

**Comment by Zireael07** [comment]
#2841, though, has much more support (even though people are clearly divided on it, with almost as many downvotes as upvotes) than this one has, probably because it doesn't involve a complete redesign of how Godot's UI nodes work (and therefore is much more likely to happen). This one feels like trying to shoehorn Godot into how CSS works (and CSS is known to have many faults and unintuitive quirks - I am a web dev by day) 

BTW I can see why https://github.com/godotengine/godot-proposals/issues/812#issuecomment-879587321 feels like "bickering", but I don't see any "cursing"

>  I'm not reviving this issue, every now and then I get an email notification that somebody commented on it and since not much have changed since the last time, I chip in with my disappointment. 

This is literally what @pycbouh means - you keep reviving the issue every time you "chip in with your disappointment". This is not very constructive since the Godot team has already seen your previous comments, they know what you don't like about the current situation

**Comment by KoBeWi** [comment]
btw if you don't like email notifications, you can unsubscribe from the issue.

**Comment by pycbouh** [comment]
Coincidentally #2841 was the exact issue I mentioned in my first comment here a month ago, as an example of us working on UX rather than redesigning the whole UI system. In fact, let's close this one as not getting traction and lacking constructive discussion in favor of #2841.

**Comment by trommlbomml** [comment]
Sorry for answering a closed issue, but I also was one of the answering people having a great interest and stopped discussing because it seems like the core team does not seem to see the benefit - or I did not follow every answer, sorry.

> I don't see many thumbs-up on the proposal and I only see what, three people in favor of it in this proposal thread.
> No one is refusing to listen to your reasons (this proposal is still open) but I simply don't see as much of a support for this as you seem to think there is (see this nice website https://godot-proposals-viewer.github.io/ displaying proposals in a rough popularity order).

Not many people (also me) know, that these thumbs are the indicator to push this feature to a priority. I think having so many comments in this thread is also an indicator, isn't it? A lot of people even do not write or thumb up but agree on many features but want to have them, so I think you make it yourself  a little bit too easy. I also look at many features and I think they make sense, but do not vote or add a thumb up.

> Most proposals, especially very complex ones, get done because there's clear, overwhelming support for them AND there is a contributor on hand who can do it. Again, just by glancing at the list linked, there are many proposals with overwhelming support that haven't happened yet because of the limited amount of contributors...

I understand that there is no high priority because you can workaround it easy sure and there are more important things. The main discussion was that the "margin concept" is kind of standard in multiple ui frameworks and is such a well-known feature seems to be for me like a great addition, but it is refused to be implemented in general.

**Comment by pycbouh** [comment]
> Not many people (also me) know, that these thumbs are the indicator to push this feature to a priority.

Thumb ups and thumb downs are the indicator for user support, not for priority. Comments can also indicate support, but we don't allow empty comments that don't add anything to the discussion, and the comments that do cannot always be evaluated as easily as a reaction on the OP. With all that said, most popular proposals don't have any issue getting thumbs on them, so even if some people miss that mechanic, it's not the majority.

We can of course explain it better in the README.

> I think having so many comments in this thread is also an indicator, isn't it?

As Zireael mentioned, most comments were made by the same couple of people, so no, unfortunately it's not really indicative of anything.

> A lot of people even do not write or thumb up but agree on many features but want to have them, so I think you make it yourself a little bit too easy. I also look at many features and I think they make sense, but do not vote or add a thumb up.

Getting a popular vote is the whole point of this repository. There are no hard rules that you need to get X number of votes for a feature to be implemented (it can be less, or it can be declined even if there is a hundred of upvotes). But the OP is still required to demonstrate the popularity. If they can do it in any other way (show a Reddit thread with a lot of upvotes for example) they can refer to it as well. But lacking any other means, users votes are what shows if there is or isn't any interest, ultimately.

If you feel like the whole process is not properly explained, feel free to open a meta proposal/a PR against this repo to amend that.

> The main discussion was that the "margin concept" is kind of standard in multiple ui frameworks and is such a well-known feature seems to be for me like a great addition, but it is refused to be implemented in general.

Once again, there are margins in Controls, several different ones in fact, depending on what you need. You have margin positioning, you have MarginContainer for insetting and styleboxes have content and expand margins for adjusting the visual positioning of control's content.

What is refused here is to implement a new independent system to replace Godot's UI system. We don't see the point in it, but we do see the point in making the UX of working on controls better. The lack of discoverability is not the same as the lack of features.

**Comment by amp-html195** [comment]
@Calinou https://lnkr.fun/4e7444d
